File: functionality/functionality_signature.py
import inspect
from typing import List, Dict, Any, Callable, Optional, Type
from dataclasses import dataclass

from util.data_type import DataType
from functionality.functionality_method import FunctionalityMethod

@dataclass
class ParameterSignature:
    """Signature for a functionality parameter"""
    name: str
    type: DataType
    
    def __init__(self, name: str, type_: DataType):
        self.name = name
        self.type = type_

class FunctionalitySignature:
    """Signature for a functionality method"""
    
    def __init__(self, name: str, parameters: List[ParameterSignature], description: str):
        self.name = name
        self.parameters = parameters
        self.description = description
    
    @classmethod
    def from_method_with_name(cls, functionality_name: str, method: Callable) -> 'FunctionalitySignature':
        """Create a signature from a method with a specified name"""
        signature = cls.from_method(method)
        signature.name = functionality_name
        return signature
    
    @classmethod
    def from_method(cls, method: Callable) -> 'FunctionalitySignature':
        """Create a signature from a method"""
        # Get the method's signature using inspect
        sig = inspect.signature(method)
        
        # Get functionality metadata if available
        functionality_name = FunctionalityMethod.get_functionality_name(method)
        description = getattr(method, 'functionality_description', "")
        
        # Create parameter signatures
        parameters = []
        for param_name, param in sig.parameters.items():
            # Skip 'self' parameter
            if param_name == 'self':
                continue
                
            # Get parameter type (use Any as default if not specified)
            param_type = param.annotation if param.annotation != inspect.Parameter.empty else Any
            data_type = DataType.from_class(param_type)
            
            parameters.append(ParameterSignature(param_name, data_type))
        
        return cls(functionality_name, parameters, description)
    
    @staticmethod
    def get_functionality_name(method: Callable) -> str:
        """Get the functionality name from a method"""
        return FunctionalityMethod.get_functionality_name(method)
    
    def get_name(self) -> str:
        """Get the name of this functionality"""
        return self.name
    
    def get_parameters(self) -> List[ParameterSignature]:
        """Get the parameters of this functionality"""
        return self.parameters
    
    def get_description(self) -> str:
        """Get the description of this functionality"""
        return self.description

File: lapras-middleware/src/main/java/kr/ac/kaist/cdsn/lapras/functionality/FunctionalitySignature.java
package kr.ac.kaist.cdsn.lapras.functionality;

import kr.ac.kaist.cdsn.lapras.util.DataType;
import org.apache.commons.lang3.text.WordUtils;

import java.lang.reflect.Method;
import java.lang.reflect.Parameter;
import java.util.ArrayList;
import java.util.List;

/**
 * Created by Daekeun Lee on 2016-12-27.
 */
public class FunctionalitySignature {
    private String name;
    private List<ParameterSignature> parameters;
    private String description;

    public static class ParameterSignature {
        private String name;
        private DataType type;

        public ParameterSignature(String name, DataType type) {
            this.name = name;
            this.type = type;
        }

        public String getName() {
            return name;
        }

        public DataType getType() {
            return type;
        }
    }

    public FunctionalitySignature(String name, List<ParameterSignature> parameters, String description) {
        this.name = name;
        this.parameters = parameters;
        this.description = description;
    }

    public FunctionalitySignature() {
    }

    public static FunctionalitySignature fromMethodWithName(String functionalityName, Method method) {
        FunctionalitySignature signature = fromMethod(method);
        signature.name = functionalityName;
        return signature;
    }

    public static FunctionalitySignature fromMethod(Method method) {
        FunctionalityMethod annotation = method.getAnnotation(FunctionalityMethod.class);
        List<ParameterSignature> arguments = new ArrayList<>(method.getParameterCount());
        for (Parameter parameter : method.getParameters()) {
            arguments.add(new ParameterSignature(parameter.getName(), DataType.fromClass(parameter.getType())));
        }

        return new FunctionalitySignature(getFunctionalityName(method), arguments, (annotation == null) ? "" : annotation.description());
    }

    public static String getFunctionalityName(Method method) {
        FunctionalityMethod functionalityMethodAnnotation = method.getAnnotation(FunctionalityMethod.class);
        if(functionalityMethodAnnotation == null) {
            return WordUtils.capitalize(method.getName());
        }
        String functionalityName = functionalityMethodAnnotation.name();
        if(functionalityName.isEmpty()) {
            functionalityName = WordUtils.capitalize(method.getName());
        }
        return functionalityName;
    }

    public String getName() {
        return name;
    }

    public List<ParameterSignature> getParameters() {
        return parameters;
    }

    public String getDescription() {
        return description;
    }
}
